presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java

```diff
@@ -0,0 +1,391 @@
+    }
+
+    private void populateColumns(JsonNode propertiesNode, List<String> arrays, List<ElasticsearchColumn> columns)
+    {
+        Comparator<String> stringComparator = (key1, key2) -> {
+            int length1 = key1.split("\\.").length;
+            int length2 = key2.split("\\.").length;
+            if (length1 == length2) {
+                return key1.compareTo(key2);
+            }
+            return length2 - length1;
+        };
+        TreeMap<String, Type> fieldsMap = new TreeMap<>(stringComparator);
+        for (String columnMetadata : getColumnMetadata(null, propertiesNode)) {
+            int delimiterIndex = columnMetadata.lastIndexOf(":");
+            if (delimiterIndex == -1 || delimiterIndex == columnMetadata.length() - 1) {
+                LOG.debug("Invalid column path format: %s", columnMetadata);
+                continue;
+            }
+            String fieldName = columnMetadata.substring(0, delimiterIndex);
+            String typeName = columnMetadata.substring(delimiterIndex + 1);
+
+            if (!fieldName.endsWith(".type")) {
+                LOG.debug("Ignoring column with no type info: %s", columnMetadata);
+                continue;
+            }
+            String propertyName = fieldName.substring(0, fieldName.lastIndexOf('.'));
+            String nestedName = propertyName.replaceAll("properties\\.", "");
+            if (nestedName.contains(".")) {
+                fieldsMap.put(nestedName, getPrestoType(typeName));
+            }
+            else {
+                boolean newNestedColumn = columns.stream()
+                        .noneMatch(column -> column.getName().equalsIgnoreCase(nestedName));
+                if (newNestedColumn) {
+                    columns.add(new ElasticsearchColumn(nestedName, getPrestoType(typeName), nestedName, typeName, arrays.contains(nestedName), -1));
+                }
+            }
+        }
+        processNestedFields(fieldsMap, columns, arrays);
+    }
+
+    private void processNestedFields(TreeMap<String, Type> fieldsMap, List<ElasticsearchColumn> columns, List<String> arrays)
+    {
+        if (fieldsMap.size() == 0) {
+            return;
+        }
+        Entry<String, Type> first = fieldsMap.firstEntry();
+        String field = first.getKey();
+        Type type = first.getValue();
+        if (field.contains(".")) {
+            String prefix = field.substring(0, field.lastIndexOf('.'));
+            ImmutableList.Builder<RowType.Field> fieldsBuilder = ImmutableList.builder();
+            int size = field.split("\\.").length;
+            Iterator<String> iterator = fieldsMap.navigableKeySet().iterator();
+            while (iterator.hasNext()) {
+                String name = iterator.next();
+                if (name.split("\\.").length == size && name.startsWith(prefix)) {
+                    fieldsBuilder.add(new RowType.Field(Optional.of(name.substring(name.lastIndexOf('.') + 1)), fieldsMap.get(name)));
```

Comment:
This is a pretty complicated expression. Please simplify (you can extract some variables here).

Reply:
will do